src/Cabal2JSON.hs

```diff
@@ -354,34 +404,72 @@ instance HasCodec Mixin where
         <*> requiredField' "include-renaming" .= mixinIncludeRenaming
 
 instance HasCodec LegacyExeDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "LegacyExeDependency" $
+      LegacyExeDependency
+        <$> requiredField' "package-name" .= (\(LegacyExeDependency s _) -> s)
+        <*> requiredField' "include-renaming" .= (\(LegacyExeDependency _ vr) -> vr)
 
 instance HasCodec ExeDependency where
   codec =
     object "ExeDependency" $
       ExeDependency
-        <$> requiredField' "package-name" .= (\(ExeDependency a b c) -> a)
-        <*> requiredField' "executable-component-name" .= (\(ExeDependency a b c) -> b)
-        <*> requiredField' "version-range" .= (\(ExeDependency a b c) -> c)
+        <$> requiredField' "package-name" .= (\(ExeDependency a _ _) -> a)
+        <*> requiredField' "executable-component-name" .= (\(ExeDependency _ b _) -> b)
+        <*> requiredField' "version-range" .= (\(ExeDependency _ _ c) -> c)
 
 instance HasCodec PkgconfigDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "PkgconfigDependency" $
+      PkgconfigDependency
+        <$> requiredField' "pkgconfigName" .= (\(PkgconfigDependency name _) -> name)
+        <*> requiredField' "pkgconfigVersionRange" .= (\(PkgconfigDependency _ version) -> version)
+
+instance HasCodec PkgconfigName where
+  codec = dimapCodec mkPkgconfigName unPkgconfigName codec
+
+instance HasCodec PkgconfigVersionRange where
+  codec = dimapCodec f g codec
+    where
+      f = \s -> case eitherParsec s of
+        Left pe -> error "Invalid version string"
+        Right a -> a
+      g = prettyShow
+
+instance HasCodec PkgconfigVersion where
+  codec = dimapCodec (PkgconfigVersion . BSU.fromString) (\(PkgconfigVersion v) -> show v) codec
```

Comment:
you shouldn't need `BSU` here. `Data.Text.decodeUtf8'` does what you want. In fact this will probably crash on invalid UTF-8 and you'll need `bimapCodec` instead.

Reply:
I feel like I have it the wrong way around because the `either` comes from the decoding but `bimapCodec` expects it to be in the encoding.

```
instance HasCodec PkgconfigVersion where
  codec = bimapCodec
    (Right . PkgconfigVersion . encodeUtf8)
    (\(PkgconfigVersion v) -> let decoded = decodeUtf8' v in case decoded of
        Left ex -> pack $ show ex
        Right txt -> txt
    )
    codec
```
What a m I doing wrong?

Reply:
> What am I doing wrong?

Ah, nothing. It's just that cabal uses `ByteString`, which cannot be acurately represented in JSON.
So what you were doing is probably fine. We still don't need `BSU`, we can use the partial `Data.Text.decodeUtf8` instead but that's a minor comment.